tools/coco_eval.py

```diff
@@ -20,8 +20,11 @@ def main():
         nargs='+',
         default=[100, 300, 1000],
         help='proposal numbers, only used for recall evaluation')
+    parser.add_argument(
+        '--class_wise', action='store_true', help='whether eval class wise ap')
```

Comment:
We can omit the underscore and just use `classwise`.

Reply:
> I ran your program to evaluate mAP of each class of [faster r50](https://s3.ap-northeast-2.amazonaws.com/open-mmlab/mmdetection/models/faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth), and compare it with the result of [codalab](https://competitions.codalab.org/competitions/20794#participate). However, the result of two are not the same. For example, this is the result of `person` category of two:
> Your program:
> 
> ```
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.522
>  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.835
>  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.554
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.366
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.598
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.667
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.180
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.523
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.598
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.451
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.663
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.740
> ```
> 
> Codalab:
> 
> ```
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.512
>  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.816
>  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.545
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.347
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.590
>  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.663
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.180
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.523
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.598
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.451
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.663
>  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.740
> ```
> 
> Please have a check about that to ensure the evaluation program is logic correct.

Thanks for the check.

I checked the code but cannot find the problem, so I just changed to use another implementation, put it into the `coco_eval` function in the `core/evaluation/coco_utils.py`, and leave the scripts in tools clean.

Since the implementation is different than before, and I compared the same JSON file in the codalab, I think this bug is fixed. But for safety, could you also check the results with the most recent commit?

Reply:
I have checked again and the results are consistent now.

Reply:
OK, thank you~